### PR TITLE
don't trust all.vars when its arguments contains get/eval

### LINKED ISSUE #4873: get() in by and the inconsistent error
```R
IDT <- data.table(iris)
vr <- "Species"
IDT[, virginca := get(vr) == "virginica"]
IDT[(virginca)][, .N, by = .(round(Sepal.Width), k = round(Sepal.Length), kar = get(vr))] 
# Works fine

IDT[(virginca), .N, by = .(round(Sepal.Width), k = round(Sepal.Length), kar = get(vr))] 
# Error in get(vr) : object 'Species' not found
```
This is the closest I could get to reproducing the error I got on data in the wild when I added a logical in i, but my actual error message was

```
column or expression 3 of 'by' or 'keyby' is type builtin. Do not quote column names. Usage: DT[,sum(colC),by=list(colA,month(colB))]
```

I'm using `R 4.0.2` and `data.table_1.13.4` on Windows 10.

### LINKED ISSUE #4981: `by` sometimes fails when it invokes `get`
```r
> tmp <- data.table(a=c(1,2), b=c(3,4), c=c(1,0))
> tmp[,.(suma=sum(a)), keyby=.(b=get("b"),c)]   # succeeds
   b c suma
1: 3 1    1
2: 4 0    2
> tmp[2,.(suma=sum(a)), keyby=.(b=b,c)]  # succeeds
   b c suma
1: 4 0    2
> tmp[2,.(suma=sum(a)), keyby=.(b=get("b"))]   # succeeds
   b suma
1: 4    2
> tmp[2,.(suma=sum(a)), keyby=.(b=get("b"),c)]   # Fails! - when `i` is non empty and `keyby` contains an additional direct column
Error in get("b") : object 'b' not found

```

`#` `Output of sessionInfo()`
R version 3.4.0 (--)     # Reproduces on R4.0.0 too
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 18.04.5 LTS

Matrix products: default
BLAS: /opt/R-3.4.0.mkl/lib64/R/lib/libR.so
LAPACK: /opt/R-3.4.0.mkl/lib64/R/lib/libRlapack.so

locale:
 [1] LC_CTYPE=en_US.UTF-8 LC_NUMERIC=C         LC_TIME=C            LC_COLLATE=C         LC_MONETARY=C       
 [6] LC_MESSAGES=C        LC_PAPER=C           LC_NAME=C            LC_ADDRESS=C         LC_TELEPHONE=C      
[11] LC_MEASUREMENT=C     LC_IDENTIFICATION=C 

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] data.table_1.14.1

loaded via a namespace (and not attached):
[1] compiler_3.4.0 tools_3.4.0

### PR DESCRIPTION
**Pull request by OfekShilon**
Closes #4873
Closes #4981
There are at least 2 other places in `[.data.table` that test for inclusion of `get` et.al. in an expression - I think they should use rapply too, but didn't touch it.

### CODE REVIEW

**Comment by OfekShilon** [comment]
Note this solves #4873 too.

**Comment by mattdowle** [comment]
news item and addition to contributor list in 531be3705290c98562530b64994b15777dc6c08d

**Comment by mattdowle** [comment]
@OfekShilon Many thanks. I invited you to be project member so amongst other things you can create branches in main repo in future. The invite is a button in your profile or projects page in github that you have to click to accept. Thanks again and welcome!

**Comment by jangorecki** [comment]
There were no unit test for the second closed issue so I added them in follow up PR https://github.com/Rdatatable/data.table/pull/4985

**Comment by OfekShilon** [comment]
@mattdowle  Thank you very much, I'll do my best to contribute. If I understand correctly - I'm not allowed to merge PRs,  right?  Also, what is the right place to ask such questions? (probably not in a PR comments)

**Comment by mattdowle** [comment]
Just me (and Arun) who can merge currently, but that may expand to Jan and Michael in future.
Yes in PRs is fine to ask such questions. We like everything in public on GitHub wherever possible. The other option would be email.

File: R/data.table.R [hunk 1]
```diff
@@ -756,7 +756,14 @@ replace_dot_alias = function(e) {
           bysub = parse(text=paste0("list(",paste(bysub,collapse=","),")"))[[1L]]
           bysubl = as.list.default(bysub)
         }
-        allbyvars = intersect(all.vars(bysub), names_x)
+        # Fix 4981: when the 'by' expression includes get/mget/eval, all.vars
+        # cannot be trusted to infer all used columns
+        bysub.elems <- rapply(as.list(bysub), as.character)
```
Comment (mattdowle) [inline R/data.table.R:761]: This is failing R 3.1.0 with : 
```
> af[, mean(a), by = "itime"]
178Error in .Primitive("as.character")(list, ...) : 
179  cannot coerce type 'builtin' to vector of type 'character'
180Calls: [ -> [.data.table -> rapply
181Execution halted
```
will fix ...

**Pull request merged**